hazelcast/src/main/java/com/hazelcast/nio/tcp/SingleProtocolDecoder.java

```diff
@@ -76,6 +76,8 @@ public HandlerStatus onRead() {
             if (protocol.equals(supportedProtocol.getDescriptor())) {
                 // replace this handler with the next one
                 channel.inboundPipeline().replace(this, inboundHandlers);
+            } else {
+                throw new IllegalStateException("Unsupported protocol exchange detected");
```

Comment:
minor: we could indicate the `protocol` that was received in the exception message.

Reply:
I print the expected protocol rather than what was received. I was a bit skeptical about printing to the output stream invalid or malicious bytes (although, i couldn't think of a way that could harm, given the size of it, but...)